controllers/utils.go

```diff
@@ -56,100 +52,128 @@ func getCDAPServiceSpec(master *v1alpha1.CDAPMaster, service ServiceName) (*v1al
 // Fail if any of the following occurs
 // - unable to find the field for the service
 // - unable to cast to CDAPStatefulServiceSpec type
 func getCDAPStatefulServiceSpec(master *v1alpha1.CDAPMaster, service ServiceName) (*v1alpha1.CDAPStatefulServiceSpec, error) {
-	val := reflect.Indirect(reflect.ValueOf(&master.Spec)).FieldByName(service)
-	if !val.IsValid() {
-		return nil, fmt.Errorf("failed to find field %v in %v", service, reflect.TypeOf(master.Spec).Name())
-	}
-	// For optional service, its service field is a pointer to spec
-	if val.Kind() == reflect.Ptr {
-		// Return nil if optional service is disabled (e.g. service field is nil)
-		if val.IsNil() {
-			return nil, nil
-		}
-	} else {
-		val = val.Addr()
-	}
-	val = reflect.Indirect(reflect.ValueOf(val.Interface())).FieldByName(fieldNameCDAPStatefulServiceSpec)
-	if !val.IsValid() {
+	if spec, err := getCDAPMasterSpec(master, service, reflect.TypeOf(v1alpha1.CDAPStatefulServiceSpec{}), true); err != nil {
+		return nil, err
+	} else if spec == nil {
 		return nil, nil
+	} else {
+		// The cast should succeed as it has been checked in the getCDAPMasterSpec method
+		return spec.(*v1alpha1.CDAPStatefulServiceSpec), nil
 	}
-	ret, ok := val.Addr().Interface().(*v1alpha1.CDAPStatefulServiceSpec)
-	if !ok {
-		return nil, fmt.Errorf("failed to cast to poiter to %v for %v", fieldNameCDAPStatefulServiceSpec, service)
-	}
-	return ret, nil
 }
 
 // Return pointer to CDAPExternalServiceSpec for the given service (using reflect) if it contains one, otherwise nil
 // Fail if any of the following occurs
 // - unable to find the field for the service
 // - unable to cast to CDAPExternalServiceSpec type
 func getCDAPExternalServiceSpec(master *v1alpha1.CDAPMaster, service ServiceName) (*v1alpha1.CDAPExternalServiceSpec, error) {
-	val := reflect.Indirect(reflect.ValueOf(&master.Spec)).FieldByName(service)
-	if !val.IsValid() {
-		return nil, fmt.Errorf("failed to find field %v in %v", service, reflect.TypeOf(master.Spec).Name())
-	}
-	val = reflect.Indirect(reflect.ValueOf(val.Addr().Interface())).FieldByName(fieldNameCDAPExternalServiceSpec)
-	if !val.IsValid() {
+	if spec, err := getCDAPMasterSpec(master, service, reflect.TypeOf(v1alpha1.CDAPExternalServiceSpec{}), true); err != nil {
+		return nil, err
+	} else if spec == nil {
 		return nil, nil
+	} else {
+		// The cast should succeed as it has been checked in the getCDAPMasterSpec method
+		return spec.(*v1alpha1.CDAPExternalServiceSpec), nil
 	}
-	ret, ok := val.Addr().Interface().(*v1alpha1.CDAPExternalServiceSpec)
-	if !ok {
-		return nil, fmt.Errorf("failed to cast to poiter to %v for %v", fieldNameCDAPExternalServiceSpec, service)
-	}
-	return ret, nil
 }
 
-func min(x, y int64) int64 {
-	if x < y {
-		return x
+// getFieldValue finds a field value in the given struct object that passes the predicate and return it as a *reflect.Value
```

Comment:
nit: the comment should call out the field to be found can be in nested structure and only the value of the first matching field name will be returned.

Reply:
Updated